### PR TITLE
Trike reliability updates

### PR DESCRIPTION
**Pull request by paulswartz**
#### Summary of changes
- Req for HTTP requests
- `active: :once` TCP mode
- batched Kinesis writes
- `SequenceNumberForOrdering` when writing

Asana: https://app.asana.com/0/584764604969369/1203329880050199/f

Notion: https://www.notion.so/mbta-downtown-crossing/Trike-Reliability-Improvements-Q4-2022-86b31cd97bec43908c672e156c98e2d6?d=dc522d42d35a4054906e4b4ddf8825ec#26ec6dd6155440ba8657137e52f4f8c0

#### Reviewer Checklist
- [x] Any new or changed functions have typespecs
- [x] Tests were added for any new functionality (don't just rely on Lcov)
- [x] Meets ticket's acceptance criteria

### CODE REVIEW

File: lib/trike/proxy.ex [hunk 1]
```diff
@@ -41,28 +51,31 @@ defmodule Trike.Proxy do
...
        clock: clock
      }, {:continue, {ref, transport}}}
   end
 
   @impl GenServer
   def handle_continue({ref, transport}, state) do
     {:ok, socket} = :ranch.handshake(ref)
-    :ok = transport.setopts(socket, active: true)
+    Logger.metadata(socket: inspect(socket))
+    :ok = transport.setopts(socket, active: :once, buffer: 131_072)
```
Comment (bklebe) [inline lib/trike/proxy.ex:69]: question: I found the docs for buffer size in `inet` but I'm still not sure which buffer it is: https://www.erlang.org/doc/man/inet.html#setopts-2. Is this the OS-level socket buffer? Some sort of internal Erlang buffer? It says it's different from `recbuf` which is for "Kernel socket buffers" How did you decide to make it this size (looks like 131KB just about)?

File: lib/trike/proxy.ex [hunk 2]
```diff
@@ -77,22 +90,49 @@ defmodule Trike.Proxy do
...
+        partition_key,
+        encoded,
+        opts
+      ])
+
+    Logger.info(
+      "put_record_timing stream=#{stream} pkey=#{inspect(partition_key)} length=#{length(records)} size=#{byte_size(encoded)} msec=#{div(usec, 1000)} result=#{result_key}"
+    )
+
+    {:ok, %{"SequenceNumber" => last_sequence_number}} = result
```
Comment (bklebe) [inline lib/trike/proxy.ex:123]: suggestion (non-blocking): this handler function is getting a bit long, perhaps we should consider extracting 102-123 into a function that takes our list of records and returns the next sequence number (or an error, or exception)?

File: lib/trike/proxy.ex [hunk 3]
```diff
@@ -77,22 +90,49 @@ defmodule Trike.Proxy do
...
+
+    state.transport.setopts(state.socket, active: :once)
```
Comment (bklebe) [inline lib/trike/proxy.ex:125]: praise: this is such a good idea! I didn't realize we could set `active: :once` more than once to keep receiving one buffer at a time, this seems much better in terms of preserving flow control on the underlying socket.

File: lib/trike/proxy.ex [hunk 4]
```diff
@@ -77,22 +90,49 @@ defmodule Trike.Proxy do
...
 
-    messages
-    |> Enum.map(&CloudEvent.from_ocs_message(&1, current_time, partition_key))
-    |> Enum.each(fn event ->
-      case Jason.encode(event) do
-        {:ok, event_json} ->
-          {:ok, _result} = state.put_record_fn.(stream, partition_key, event_json)
+    records =
+      messages
+      |> Enum.map(&CloudEvent.from_ocs_message(&1, current_time, partition_key))
```
Comment (bklebe) [inline lib/trike/proxy.ex:100]: question: This seems like the most straightforward, least-overhead way of encoding multiple messages, but it does mean that our CloudEvent format doesn't describe the structure of our Kinesis records anymore. This doesn't seem like a big deal, but have you considered either updating our CloudEvent format or defining a new one for encasing multiple messages in a record? We could also just make that a ticket and push it off.

File: lib/trike/proxy.ex [hunk 5]
```diff
@@ -77,22 +90,49 @@ defmodule Trike.Proxy do
...
+    {usec, {result_key, _} = result} =
+      :timer.tc(state.put_record_fn, [
+        stream,
+        partition_key,
+        encoded,
+        opts
+      ])
+
+    Logger.info(
+      "put_record_timing stream=#{stream} pkey=#{inspect(partition_key)} length=#{length(records)} size=#{byte_size(encoded)} msec=#{div(usec, 1000)} result=#{result_key}"
```
Comment (bklebe) [inline lib/trike/proxy.ex:120]: question: what's the thinking behind this new log message? Making it easier to observe Trike performance in Splunk?

File: lib/trike/proxy.ex [hunk 6]
```diff
@@ -111,6 +151,13 @@ defmodule Trike.Proxy do
     {:noreply, state}
   end
 
+  @impl GenServer
+  def terminate(_reason, state) do
+    Logger.info("Terminating")
+    state.transport.close(state.socket)
+    {:ok, state}
+  end
```
[the remark this answers is not in the document]
Comment (paulswartz) [reply lib/trike/proxy.ex:159]: We do, but as a part of the metadata:
- https://github.com/mbta/trike/pull/20/files#diff-8e72c709985cb6a43cc0ab32cb487b8c17d2b2d003338ea42a4e2ad2829979dcR10 adds it to the logger configuration
- https://github.com/mbta/trike/pull/20/files#diff-1665f8f11d0ff3389e858a42159b39d8468e8ae67279547e52ff4ba76851aa52R68 sets the metadata after the handshake

These ensure that all messages from a given Proxy instance are tagged with their PID and the socket that's connected.

File: lib/trike/proxy.ex (hunk 1 above)
Comment (paulswartz) [reply lib/trike/proxy.ex:69]: It's 128KB. I believe this is an VM-internal buffer, as it's the maximum size of `data` that we receive. TBH, it's probably overkill as the largest packets I saw in practice were about 36K, but even 128K over 4 sockets isn't much extra memory.

File: lib/trike/proxy.ex (hunk 2 above)
Comment (paulswartz) [reply lib/trike/proxy.ex:123]: Great suggestion!

[ ] refactor `handle_info` to do less

File: lib/trike/proxy.ex (hunk 3 above)
Comment (paulswartz) [reply lib/trike/proxy.ex:125]: This is the pattern I've seen in other applications. If you don't re-set the `active: :once` when you're ready, you don't get another packet (as a message, at least).

File: lib/trike/proxy.ex (hunk 4 above)
Comment (paulswartz) [reply lib/trike/proxy.ex:100]: I followed the CloudEvent spec for batching JSON records (https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/formats/json-format.md#4-json-batch-format) so I think this is still compatible.

File: lib/trike/proxy.ex (hunk 5 above)
Comment (paulswartz) [reply lib/trike/proxy.ex:120]: Yep, exactly! This shows how much time was spent sending to the records to Kinesis, along with the size of the payload and the number of records.

**Review by bklebe — APPROVED** [review]

**Pull request merged**